changelog/@unreleased/pr-4879.v2.yml

```diff
@@ -0,0 +1,40 @@
+type: improvement
+improvement:
+  description: |-
+    add refresh failure callbacks
+
+    **Goals (and why)**:
+
+    Adds an option to register callback functions that get called whenever a
+    LockRefreshingLockService stops refreshing a lock because it failed to
+    refresh.
+
+    This helps with situations where a client intentionally will be holding
+    locks for a long time, and they want to be nofied immediately as soon as
+    refreshing the lock fails. Without this callback, clients would need to
+    implement their own background thread to check if the locks were still
+    valid.
+
+    We have an internal product that would already like to use this feature, and this will save us an additional background task that would currently be used to repeatedly check that locks are still valid.
+
+    **Implementation Description (bullets)**:
+
+    Adds methods to register and remove callbacks from LockRefreshingLockService, and calls the callback when lock tokens fail to refresh. To avoid problems where a callback is very slow, the callbacks are only called on a new thread that is submitted to the same executor that runs refreshes. The executor service already has an unbounded max number of threads so it will create new threads and for these callbacks and then dispose of the threads when not needed.
```

Comment:
> The executor service already has an unbounded max number of threads

I think it's a single threaded scheduled executor not an unbounded executor, right?

Reply:
It's a ScheduledThreadPoolExecutor with a corePoolSize of 1. If you follow through to when it actually constructs it the maxPoolSize is Integer.MAX_VALUE.

Reply:
Executors are subtle and dangerous creatures. Can you write a test that shows tasks executing in parallel on this executor?

Reply:
You're right, this doesn't appear to work as I had expected right now. I think I'll need to use a separate ExecutorService for the callbacks and use a SynchronousQueue so that it actually creates new threads on demand instead of just queuing up tasks.

Reply:
Added a separate cached executor service to use for the callbacks so this should work as expected now (and a test that asserts it actually does run the callbacks in parallel).